astropy/coordinates/baseframe.py

```diff
@@ -424,39 +424,121 @@ def __init__(self, *args, copy=True, representation_type=None,
         representation_type = kwargs.pop('representation_type', representation_type)
 
         if representation_type is not None or differential_type is not None:
+            representation_type, differential_type = self._infer_representation_differential_type(
+                representation_type, differential_type)
+            self.set_representation_cls(representation_type, **differential_type)
+        else:
+            self._representation = {'base': self.default_representation,
+                                    's': self.default_differential}
 
-            if representation_type is None:
-                representation_type = self.default_representation
+        representation_data, differential_data = self._infer_representation_differential_data(
+            args, copy, kwargs)
+
+        if differential_data:
+            self._data = representation_data.with_differentials(
+                {'s': differential_data})
+        else:
+            self._data = representation_data  # possibly None.
 
-            if (inspect.isclass(differential_type)
-                    and issubclass(differential_type, r.BaseDifferential)):
-                # TODO: assumes the differential class is for the velocity
-                # differential
-                differential_type = {'s': differential_type}
+        # Set frame attributes, if any
 
-            elif isinstance(differential_type, str):
-                # TODO: assumes the differential class is for the velocity
-                # differential
-                diff_cls = r.DIFFERENTIAL_CLASSES[differential_type]
-                differential_type = {'s': diff_cls}
+        values = {}
+        for fnm, fdefault in self.get_frame_attr_names().items():
+            # Read-only frame attributes are defined as FrameAttribute
+            # descriptors which are not settable, so set 'real' attributes as
+            # the name prefaced with an underscore.
+
+            if fnm in kwargs:
+                value = kwargs.pop(fnm)
+                setattr(self, '_' + fnm, value)
+                # Validate attribute by getting it. If the instance has data,
+                # this also checks its shape is OK. If not, we do it below.
+                values[fnm] = getattr(self, fnm)
```

Comment:
I wanted to decouple the setting of the frame attributes from the rest, but found that a lot of things are happening in this `getattr(self, fnm)`: it's where the validation of the shapes takes place _if the frame has `_data` already_ (see my extra code comment in `Attribute.__get__`). I wonder if that part could be splitted out to another function, instead of validating on the fly like this?

Reply:
Yes, this is quite ugly (I refactored it to get shape consistency before). I think it may be an idea to treat the data and values as one giant to-be-broadcasted whole instead. But my suggestion would be to leave it for another PR.

Reply:
Absolutely, let's leave this for another PR. Just touching base to make sure I was not missing anything.